### PR TITLE
feat(image): throw if alt text is missing

### PR DESCRIPTION
**Pull request by DerYeger**
## Changes

- #4291 
- `<Image>` and `<Picture>` now throw if `alt` prop is missing
- Prop types have been updated to require `alt`

## Testing

This change was tested by adding fixtures that have images without `alt` text.
The tests assert that an error is thrown.

## Docs

PR: https://github.com/withastro/docs/pull/1421

### CODE REVIEW

File: packages/integrations/image/README.md
Comment (sarah11918) [inline packages/integrations/image/README.md:96]: LOVE this new feature, @DerYeger! I have installed some Chrome extensions myself to make sure I'm not allowed to Tweet without alt-text on my images.  😄 

Please ALSO check the "Usage" heading further down the page where the attributes, and whether or not they are required are listed for both components. There should be an `alt` heading and description for each one of them. It looks like it's currently not there for `<Image />` but is there for `<Picture />` though only saying "if it's provided."

These should both prominently display the new info, and double check all the examples to make sure they all show `alt=` on every component. 

Lastly, @DerYeger  & @tony-navillus, what exactly *is* the error message that is thrown? (Sorry, hard to look through everything on my phone here) If it's not, can it be an explicit message to add alt text rather than a generic error?

Thanks!

Comment (DerYeger) [reply packages/integrations/image/README.md:96]: Hi, I used the suggestion of https://github.com/withastro/astro/issues/4291#issuecomment-1215949055 as the error message, replacing `image` with `picture` for the respective component.

**Review by tony-sull — APPROVED** [review]
This is excellent, thanks for the thorough test coverage!

Consider the feature work here 🎉  @sarah11918 I'll leave the docs approval up to you if you don't mind, once it has your approval as well I'll get this merged in and ready for release 🚀

**Comment by sarah11918** [comment]
Yay! Will look closely when I am back at a laptop!

File: packages/integrations/image/README.md
Comment (sarah11918) [reply packages/integrations/image/README.md:96]: Great! I've merged the PR for the docs "Images Guide" page just now. Just let me know if you're up for making the above edits under "Usage" (as an attribute, like all the other `aspectRatio`, `src` etc.) in addition to your note here, or whether you'd like me to take care of that!  But, it DOES need to be there as an attribute for each component.

**Comment by tony-sull** [comment]
@DerYeger Let me know if you have any issues with the merge conflict!

**Comment by tony-sull** [comment]
I'll get those tests passing, looks like the tests in main are too aggressive with hard-coded filename hashes

**Comment by sarah11918** [comment]
@tony-sull Ooh, good! I need to make a fix to this README anyway. Please do not merge this until I've updated the docs here. 💜

File: packages/integrations/image/README.md
Comment (sarah11918) [reply packages/integrations/image/README.md:96]: These changes don't appear to have been made, so I'm making them now here!

**Comment by sarah11918** [comment]
OK @tony-sull PLEASE check all my alt-text stuff carefully! Code samples! etc! Please make sure it's actually correct and no typos.  💜 Otherwise, I'm happy.

**Comment by DerYeger** [comment]
> I'll get those tests passing, looks like the tests in main are too aggressive with hard-coded filename hashes

Thanks! Let me know I can help out anywhere else.

**Comment by tony-sull** [comment]
Just opened #4579 with a big cleanup of the image tests.  This PR was a perfect kick in the pants to stop dragging my feet on refactoring all the test code 😄

File: packages/integrations/image/README.md
Comment (FredKSchott) [inline packages/integrations/image/README.md:217]: add a note here about why required, how to use `alt=""` if its decorative, etc. You can probably copy-paste this from the note in the README

**Review by FredKSchott — COMMENTED** [review]
Looks great! I had a few comments, mainly around documentation. Thanks for pushing us to add this, I'm a big +1 for this feature.

**Comment by FredKSchott** [comment]
@tony-sull @DerYeger I'm realizing that this will break most existing users who didn't add alt text, and my guess is that this is a lot of them. Even if we make this error helpful, that's still going to be a real pain to upgrade given that we won't even render the page anymore if `alt` is missing.

Thoughts on making this a TS-only change first, and then adding the runtime error in a future minor release? This will still provide a very visible error in `astro check` and visible inline errors in their code editor, but it will also give them some time to update before we escalate.

FYI to implement this change, everything about this PR can stay as-is except for removing the ~3 lines that throw the error at runtime.

**Comment by matthewp** [comment]
@FredKSchott are you proposing essentially we warn in 1 "major" (minor in this case because < 1.0) release and then break in the next? If so I think that's fine and in line with some other changes we've done.

**Comment by tony-sull** [comment]
@FredKSchott That's a good callout, I like the idea of warning for now and throwing the error in a later release (either a future minor or 1.0)

@DerYeger The tests should be *almost* there now, I refactored all the image tests to avoid using hard-coded filename hashes but it looks like I missed an `alt` tag in the `with-mdx` test suite 🤦

**Comment by tony-sull** [comment]
@FredKSchott @sarah11918 I pushed a few updates here to include the comments above.  Any additional feedback before merging this in?

For the warning, I wrapped it in a helper to make sure it's only logged once per `astro dev` or `astro build`.  Logging a warning for every component could get very frustrating 😅

**Comment by DerYeger** [comment]
> @FredKSchott @sarah11918 I pushed a few updates here to include the comments above. Any additional feedback before merging this in?
> 
> For the warning, I wrapped it in a helper to make sure it's only logged once per `astro dev` or `astro build`. Logging a warning for every component could get very frustrating 😅

That’s a great change, thanks for cleaning up this PR!

**Comment by tony-sull** [comment]
Thanks for this PR, @DerYeger!  Excellent test coverage and docs updates 🎉 

Very excited to see this flipped over to a build error in the future, I'm always a fan of build tools making sure I don't accidentally create a11y issues!

**Comment by FredKSchott** [comment]
> @FredKSchott are you proposing essentially we warn in 1 "major" (minor in this case because < 1.0) release and then break in the next? If so I think that's fine and in line with some other changes we've done.

Yup, pretty much! Glad everyone was on board for this, sgtm and excited to see this merged!

**Pull request merged**
